File: src/nova_audio_agent/realtime/history.py
"""Pure canonical conversation projection for bounded realtime recovery experiments."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Literal, Sequence

from nova_audio_agent.memory import CONVERSATION_CHANNEL, MemoryItem, Trust

MAX_PACKED_RECOVERY_CONTENT = 3900


@dataclass(frozen=True, slots=True)
class RecoveryTurn:
    sequence: int
    role: Literal["user", "assistant"]
    text: str
    delivery: Literal["user_final", "spoken"]
    played_ms: int | None
    trust: Trust
    source: Literal["conversation"] = "conversation"

    def __post_init__(self) -> None:
        if type(self.sequence) is not int or self.sequence < 1:
            raise ValueError("recovery sequence must be positive")
        if type(self.text) is not str or not self.text.strip():
            raise ValueError("recovery text must be non-empty")
        if self.source != "conversation":
            raise ValueError("recovery source must be conversation")
        if self.role == "user":
            if (
                self.delivery != "user_final"
                or self.trust != "trusted_user"
                or self.played_ms is not None
            ):
                raise ValueError("invalid canonical user recovery turn")
        elif self.role == "assistant":
            if self.delivery != "spoken" or self.trust != "trusted_system":
                raise ValueError("invalid canonical assistant recovery turn")
            if self.played_ms is not None and (
                type(self.played_ms) is not int or self.played_ms < 0
            ):
                raise ValueError("played_ms must be non-negative")
        else:
            raise ValueError("unknown recovery role")


def project_recovery_turns(
    items: Sequence[MemoryItem],
    *,
    max_pairs: int,
    max_chars: int = 3500,
) -> tuple[RecoveryTurn, ...]:
    """Return the newest whole trusted user/spoken-assistant pairs within both bounds."""
    if max_pairs <= 0 or max_chars <= 0:
        return ()
    pairs: list[tuple[RecoveryTurn, RecoveryTurn]] = []
    pending_user: RecoveryTurn | None = None
    for item in items:
        if item.channel != CONVERSATION_CHANNEL or item.outcome is not None:
            continue
        text = item.content.get("text")
        if type(text) is not str or not text.strip():
            continue
        if item.trust == "trusted_user" and "delivery" not in item.content:
            pending_user = RecoveryTurn(
                sequence=item.seq,
                role="user",
                text=text,
                delivery="user_final",
                played_ms=None,
                trust="trusted_user",
            )
            continue
        if (
            pending_user is None
            or item.trust != "trusted_system"
            or item.content.get("delivery") != "spoken"
        ):
            if pending_user is not None and item.trust == "trusted_system":
                pending_user = None
            continue
        played_ms = item.content.get("played_ms")
        if type(played_ms) is not int:
            played_ms = None
        pairs.append(
            (
                pending_user,
                RecoveryTurn(
                    sequence=item.seq,
                    role="assistant",
                    text=text,
                    delivery="spoken",
                    played_ms=played_ms,
                    trust="trusted_system",
                ),
            )
        )
        pending_user = None

    selected = pairs[-max_pairs:]
    while selected and sum(len(turn.text) for pair in selected for turn in pair) > max_chars:
        selected.pop(0)
    return tuple(turn for pair in selected for turn in pair)


def pack_recovery_turns(
    history: Sequence[RecoveryTurn],
    *,
    max_chars: int = MAX_PACKED_RECOVERY_CONTENT,
) -> tuple[tuple[RecoveryTurn, ...], str]:
    """Encode newest whole pairs without ever slicing encoded JSON."""
    turns = list(history)
    while turns:
        content = json.dumps(
            {
                "version": 1,
                "turns": [
                    {"sequence": turn.sequence, "role": turn.role, "text": turn.text}
                    for turn in turns
                ],
            },
            ensure_ascii=False,
            separators=(",", ":"),
        )
        if len(content) <= max_chars:
            return tuple(turns), content
        del turns[:2]
    return (), ""

File: runtime/src/realtime/history.ts
import { z } from 'zod'
import { CONVERSATION_CHANNEL, type MemoryItem } from '../memory.js'
import {stripLikePython} from '../python-text.js'

export const MAX_PACKED_RECOVERY_CONTENT = 3900

const recoveryTurnBase = z.object({
  sequence: z.number().int().positive(),
  text: z.string().refine(value => stripLikePython(value) !== '', 'recovery text must be non-empty'),
  source: z.literal('conversation').default('conversation'),
}).strict()

const userRecoveryTurnSchema = recoveryTurnBase.extend({
  role: z.literal('user'),
  delivery: z.literal('user_final'),
  played_ms: z.null(),
  trust: z.literal('trusted_user'),
})

const assistantRecoveryTurnSchema = recoveryTurnBase.extend({
  role: z.literal('assistant'),
  delivery: z.literal('spoken'),
  played_ms: z.number().int().nonnegative().nullable(),
  trust: z.literal('trusted_system'),
})

export const recoveryTurnSchema = z.discriminatedUnion('role', [
  userRecoveryTurnSchema,
  assistantRecoveryTurnSchema,
])

export type RecoveryTurn = z.infer<typeof recoveryTurnSchema>

export function projectRecoveryTurns(
  items: readonly MemoryItem[],
  options: {readonly maxPairs: number; readonly maxChars?: number},
): readonly RecoveryTurn[] {
  const maxChars = options.maxChars ?? 3500
  if (!Number.isInteger(options.maxPairs)) throw new TypeError('maxPairs must be an integer')
  if (options.maxPairs <= 0 || maxChars <= 0) return []

  const pairs: [RecoveryTurn, RecoveryTurn][] = []
  let pendingUser: RecoveryTurn | undefined
  for (const item of items) {
    if (item.channel !== CONVERSATION_CHANNEL || item.outcome !== null) continue
    const text = item.content.text
    if (typeof text !== 'string' || stripLikePython(text) === '') continue
    if (item.trust === 'trusted_user' && !Object.hasOwn(item.content, 'delivery')) {
      pendingUser = recoveryTurnSchema.parse({
        sequence: item.seq,
        role: 'user',
        text,
        delivery: 'user_final',
        played_ms: null,
        trust: 'trusted_user',
      })
      continue
    }
    if (
      pendingUser === undefined
      || item.trust !== 'trusted_system'
      || item.content.delivery !== 'spoken'
    ) {
      if (pendingUser !== undefined && item.trust === 'trusted_system') pendingUser = undefined
      continue
    }
    const playedMs = item.content.played_ms
    const assistant = recoveryTurnSchema.parse({
      sequence: item.seq,
      role: 'assistant',
      text,
      delivery: 'spoken',
      played_ms: typeof playedMs === 'number' && Number.isInteger(playedMs) ? playedMs : null,
      trust: 'trusted_system',
    })
    pairs.push([pendingUser, assistant])
    pendingUser = undefined
  }

  const selected = pairs.slice(-options.maxPairs)
  while (selected.length > 0 && pairTextLength(selected) > maxChars) selected.shift()
  return selected.flat()
}

export function packRecoveryTurns(
  history: readonly RecoveryTurn[],
  options: {readonly maxChars?: number} = {},
): {readonly turns: readonly RecoveryTurn[]; readonly content: string} {
  const maxChars = options.maxChars ?? MAX_PACKED_RECOVERY_CONTENT
  const turns = history.map(turn => recoveryTurnSchema.parse(turn))
  while (turns.length > 0) {
    const content = JSON.stringify({
      version: 1,
      turns: turns.map(turn => ({sequence: turn.sequence, role: turn.role, text: turn.text})),
    })
    if (codePointLength(content) <= maxChars) return {turns, content}
    turns.splice(0, 2)
  }
  return {turns: [], content: ''}
}

function pairTextLength(pairs: readonly (readonly [RecoveryTurn, RecoveryTurn])[]): number {
  return pairs.reduce((total, pair) => (
    total + codePointLength(pair[0].text) + codePointLength(pair[1].text)
  ), 0)
}

function codePointLength(value: string): number {
  return [...value].length
}
